datafusion/physical-expr/src/aggregate/average.rs

```diff
@@ -383,6 +417,234 @@ impl RowAccumulator for AvgRowAccumulator {
+                                }
+                                index_mask <<= 1;
+                            },
+                        )
+                    });
+
+                let remainder_bits = bit_chunks.remainder_bits();
+                group_indices_remainder
+                    .iter()
+                    .zip(data_remainder.iter())
+                    .enumerate()
+                    .for_each(|(i, (group_index, new_value))| {
+                        if remainder_bits & (1 << i) != 0 {
+                            let sum = &mut self.sums[*group_index];
+                            *sum = sum.add_wrapping(*new_value);
+                        }
+                    });
+            }
+        }
+        Ok(())
+    }
+}
+
+impl<T, F> GroupsAccumulator for AvgGroupsAccumulator<T, F>
+where
+    T: ArrowNumericType + Send,
+    F: Fn(T::Native, u64) -> Result<T::Native> + Send,
+{
+    fn update_batch(
+        &mut self,
+        values: &[ArrayRef],
+        group_indicies: &[usize],
+        opt_filter: Option<&arrow_array::BooleanArray>,
+        total_num_groups: usize,
+    ) -> Result<()> {
+        assert_eq!(values.len(), 1, "single argument to update_batch");
+        let values = values.get(0).unwrap().as_primitive::<T>();
+
+        // update counts (TOD account for opt_filter)
+        self.counts.resize(total_num_groups, 0);
+        group_indicies.iter().for_each(|&group_idx| {
+            self.counts[group_idx] += 1;
+        });
+
+        // update values
+        self.update_sums(values, group_indicies, opt_filter, total_num_groups)?;
+
+        Ok(())
+    }
+
+    fn merge_batch(
+        &mut self,
+        values: &[ArrayRef],
+        group_indicies: &[usize],
+        opt_filter: Option<&arrow_array::BooleanArray>,
+        total_num_groups: usize,
+    ) -> Result<()> {
+        assert_eq!(values.len(), 2, "two arguments to merge_batch");
+        // first batch is counts, second is partial sums
+        let counts = values.get(0).unwrap().as_primitive::<UInt64Type>();
+        let partial_sums = values.get(1).unwrap().as_primitive::<T>();
+
+        // update counts by summing the partial sums (TODO account for opt_filter)
+        self.counts.resize(total_num_groups, 0);
+        group_indicies.iter().zip(counts.values().iter()).for_each(
+            |(&group_idx, &count)| {
+                self.counts[group_idx] += count;
+            },
+        );
+
+        // update values
+        self.update_sums(partial_sums, group_indicies, opt_filter, total_num_groups)?;
+
+        Ok(())
+    }
+
+    fn evaluate(&mut self) -> Result<ArrayRef> {
+        let counts = std::mem::take(&mut self.counts);
+        let sums = std::mem::take(&mut self.sums);
+
+        let averages: Vec<T::Native> = sums
+            .into_iter()
+            .zip(counts.into_iter())
+            .map(|(sum, count)| (self.avg_fn)(sum, count))
+            .collect::<Result<Vec<_>>>()?;
```

Comment:
When adding a where clause on `AvgGroupsAccumulator` with `NativeAdapter<T>: From<<T as ArrowPrimitiveType>::Native>`
we can collect it to `PrimitiveArray` instead of collecting into `Vec`

Reply:
Thanks @Dandandan ! 

I probably didn't totally understand how to use this tip. Here is  what I tried:

```diff
index ee249f3bd1..36155ada18 100644
--- a/datafusion/physical-expr/src/aggregate/average.rs
+++ b/datafusion/physical-expr/src/aggregate/average.rs
@@ -34,12 +34,12 @@ use crate::aggregate::utils::down_cast_any_ref;
 use crate::expressions::format_state_name;
 use crate::{AggregateExpr, GroupsAccumulator, PhysicalExpr};
 use arrow::compute;
-use arrow::datatypes::{DataType, Decimal128Type, UInt64Type};
+use arrow::datatypes::{DataType, Decimal128Type, UInt64Type, ArrowPrimitiveType};
 use arrow::{
     array::{ArrayRef, UInt64Array},
     datatypes::Field,
 };
-use arrow_array::{Array, ArrowNativeTypeOp, ArrowNumericType, PrimitiveArray};
+use arrow_array::{Array, ArrowNativeTypeOp, ArrowNumericType, PrimitiveArray, NativeAdapter};
 use datafusion_common::{downcast_value, ScalarValue};
 use datafusion_common::{DataFusionError, Result};
 use datafusion_expr::Accumulator;
@@ -570,6 +570,7 @@ where
 impl<T, F> GroupsAccumulator for AvgGroupsAccumulator<T, F>
 where
     T: ArrowNumericType + Send,
+    NativeAdapter<T>: From<<T as ArrowPrimitiveType>::Native>,
     F: Fn(T::Native, u64) -> Result<T::Native> + Send,
 {
     fn update_batch(
@@ -620,8 +621,8 @@ where
             .map(|(sum, count)| (self.avg_fn)(sum, count))
             .collect::<Result<Vec<_>>>()?;
 
-        // TODO figure out how to do this without the iter / copy
-        let array = PrimitiveArray::<T>::from_iter_values(averages);
+        // Create array directly from Vec:
+        let array = PrimitiveArray::<T>::from(averages);
 
         // fix up decimal precision and scale for decimals
         let array = adjust_output_array(&self.return_data_type, Arc::new(array))?;
```


But the compiler generates this error

```
error[E0277]: the trait bound `arrow_array::PrimitiveArray<T>: From<Vec<<T as arrow_array::ArrowPrimitiveType>::Native>>` is not satisfied
   --> datafusion/physical-expr/src/aggregate/average.rs:625:47
    |
625 |         let array = PrimitiveArray::<T>::from(averages);
    |                     ------------------------- ^^^^^^^^ the trait `From<Vec<<T as arrow_array::ArrowPrimitiveType>::Native>>` is not implemented for `arrow_array::PrimitiveArray<T>`
    |                     |
    |                     required by a bound introduced by this call
    |
help: consider extending the `where` clause, but there might be an alternative better way to express this requirement
    |
574 |     F: Fn(T::Native, u64) -> Result<T::Native> + Send, arrow_array::PrimitiveArray<T>: From<Vec<<T as arrow_array::ArrowPrimitiveType>::Native>>
    |                                                      ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

Reply:
I meant we can collect to the primitve array directly instead of creating a temporary `Vec`:

```
        let averages: PrimitiveArray<T::Native> = sums
            .into_iter()
            .zip(counts.into_iter())
            .map(|(sum, count)| (self.avg_fn)(sum, count))
            .collect::<Result<PrimitiveArray<_>>>()?;
```

Reply:
I tried hat
```
   Compiling datafusion-physical-expr v27.0.0 (/Users/alamb/Software/arrow-datafusion/datafusion/physical-expr)
error[E0277]: the trait bound `<T as arrow_array::ArrowPrimitiveType>::Native: arrow_array::ArrowPrimitiveType` is not satisfied
   --> datafusion/physical-expr/src/aggregate/average.rs:619:20
    |
619 |         let array: PrimitiveArray<T::Native> = sums
    |                    ^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `arrow_array::ArrowPrimitiveType` is not implemented for `<T as arrow_array::ArrowPrimitiveType>::Native`
    |
note: required by a bound in `arrow_array::PrimitiveArray`
   --> /Users/alamb/.cargo/registry/src/index.crates.io-6f17d22bba15001f/arrow-array-42.0.0/src/array/primitive_array.rs:477:30
    |
477 | pub struct PrimitiveArray<T: ArrowPrimitiveType> {
    |                              ^^^^^^^^^^^^^^^^^^ required by this bound in `PrimitiveArray`
help: consider further restricting the associated type
    |
614 |     fn evaluate(&mut self) -> Result<ArrayRef> where <T as arrow_array::ArrowPrimitiveType>::Native: arrow_array::ArrowPrimitiveType {
    |                                                +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++

error[E0277]: the trait bound `<T as arrow_array::ArrowPrimitiveType>::Native: arrow_array::ArrowPrimitiveType` is not 
```

Perhaps collecting directly into a naive vec is not supported for fallable iterators 🤔

Reply:
You have to add it to `impl<T, F> AvgGroupsAccumulator` as well ;)

Reply:
```diff
index ee249f3bd..4117e45e2 100644
--- a/datafusion/physical-expr/src/aggregate/average.rs
+++ b/datafusion/physical-expr/src/aggregate/average.rs
@@ -39,7 +39,7 @@ use arrow::{
     array::{ArrayRef, UInt64Array},
     datatypes::Field,
 };
-use arrow_array::{Array, ArrowNativeTypeOp, ArrowNumericType, PrimitiveArray};
+use arrow_array::{Array, ArrowNativeTypeOp, ArrowNumericType, PrimitiveArray, NativeAdapter};
 use datafusion_common::{downcast_value, ScalarValue};
 use datafusion_common::{DataFusionError, Result};
 use datafusion_expr::Accumulator;
@@ -428,6 +428,7 @@ struct AvgGroupsAccumulator<T, F>
 where
     T: ArrowNumericType + Send,
     F: Fn(T::Native, u64) -> Result<T::Native> + Send,
+    NativeAdapter<T>: From<<T as arrow_array::ArrowPrimitiveType>::Native>
 {
     /// The type of the internal sum
     sum_data_type: DataType,
@@ -449,6 +450,7 @@ impl<T, F> AvgGroupsAccumulator<T, F>
 where
     T: ArrowNumericType + Send,
     F: Fn(T::Native, u64) -> Result<T::Native> + Send,
+    NativeAdapter<T>: From<<T as arrow_array::ArrowPrimitiveType>::Native>
 {
     pub fn new(sum_data_type: &DataType, return_data_type: &DataType, avg_fn: F) -> Self {
         debug!(
@@ -571,6 +573,7 @@ impl<T, F> GroupsAccumulator for AvgGroupsAccumulator<T, F>
 where
     T: ArrowNumericType + Send,
     F: Fn(T::Native, u64) -> Result<T::Native> + Send,
+    NativeAdapter<T>: From<<T as arrow_array::ArrowPrimitiveType>::Native>
 {
     fn update_batch(
         &mut self,
@@ -614,14 +617,11 @@ where
         let counts = std::mem::take(&mut self.counts);
         let sums = std::mem::take(&mut self.sums);
 
-        let averages: Vec<T::Native> = sums
+        let array: PrimitiveArray<T> = sums
             .into_iter()
             .zip(counts.into_iter())
             .map(|(sum, count)| (self.avg_fn)(sum, count))
-            .collect::<Result<Vec<_>>>()?;
-
-        // TODO figure out how to do this without the iter / copy
-        let array = PrimitiveArray::<T>::from_iter_values(averages);
+            .collect::<Result<PrimitiveArray<_>>>()?;
 
         // fix up decimal precision and scale for decimals
         let array = adjust_output_array(&self.return_data_type, Arc::new(array))?;
```

Reply:
Sorry I saw it's using `PrimitiveArray<T>` instead as well :)

Reply:
Aah yes, this is caused by the lack of trait specialization forcing manual impls. The following should work and be optimal (collecting via NativeAdapter will not perform as well)

```
let averages: Vec<T::Native> = sums
             .into_iter()
             .zip(counts.into_iter())
             .map(|(sum, count)| (self.avg_fn)(sum, count))

let array = PrimitiveArray::new(averages.into(), None);
```

Reply:
Thank you. I will try these when I get time probably tomorrow. I really appreciate the help from both of you

Reply:
It worked great:
```diff
diff --git a/datafusion/physical-expr/src/aggregate/average.rs b/datafusion/physical-expr/src/aggregate/average.rs
index ee249f3bd1..8de2460a17 100644
--- a/datafusion/physical-expr/src/aggregate/average.rs
+++ b/datafusion/physical-expr/src/aggregate/average.rs
@@ -620,8 +620,9 @@ where
             .map(|(sum, count)| (self.avg_fn)(sum, count))
             .collect::<Result<Vec<_>>>()?;
 
-        // TODO figure out how to do this without the iter / copy
-        let array = PrimitiveArray::<T>::from_iter_values(averages);
+        // Create a primitive array (without a copy)
+        let nulls = None; // TODO implement null handling
+        let array = PrimitiveArray::<T>::new(averages.into(), nulls);
 
         // fix up decimal precision and scale for decimals
         let array = adjust_output_array(&self.return_data_type, Arc::new(array))?;
@@ -637,8 +638,8 @@ where
 
         let sums = std::mem::take(&mut self.sums);
         // create array from vec is zero copy
-        // TODO figure out how to do this without the iter / copy
-        let sums: PrimitiveArray<T> = PrimitiveArray::from_iter_values(sums);
+        let nulls = None; // TODO implement null handling
+        let sums = PrimitiveArray::<T>::new(sums.into(), nulls);
 
         // fix up decimal precision and scale for decimals
         let sums = adjust_output_array(&self.sum_data_type, Arc::new(sums))?;
```

👍